sprs/src/io.rs

```diff
@@ -134,9 +143,15 @@ where
     reader.read_line(&mut line)?;
     let header = line.to_lowercase();
     let (sym_mode, data_type) = parse_header(&header)?;
-    if data_type == DataType::Complex {
-        // we currently don't support complex
-        return Err(UnsupportedMatrixMarketFormat);
+    let data_num_kind = match data_type {
+        DataType::Integer => NumKind::Integer,
+        DataType::Real => NumKind::Float,
+        DataType::Complex => NumKind::Complex,
+    };
+    if (N::num_kind() == NumKind::Complex || data_num_kind == NumKind::Complex)
```

Comment:
We can simplify parsing a bit if we require the user to do the conversion themselves, maybe this is a less suprising behaviour than trying to do implicit conversion?

Reply:
In `int_matrix_market_read`, the file `simple_int.mm` is read into an `i32` matrix.  I think you're saying that this test should be changed so it reads from an MM file that has an "integer" in the header ?  If so, I will proceed on this change (and the other related changes).

This will mean that instead of that convoluted logic comparing the data-type and matrix num_kinds, we will just check for strict equality of the two kinds.

I think this is what you mean, yes?

Reply:
Yes, just allowing int -> int, float -> float, complex -> complex. Otherwise we need to support a conversion path from int -> complex

Reply:
Hmm, or we support this by changing the format test to
```rust
    match (data_num_kind, N::num_kind()) {
        (NumKind::Integer, _)
        | (NumKind::Float, NumKind::Float | NumKind::Complex)
        | (NumKind::Complex, NumKind::Complex) => {}
        _ => return Err(WrongReadFormat),
    }
```
But then we need to make sure we can read an int matrix market into a complex type

Reply:
Two thoughts:

1. Scipy does not support this: the read operation on an MM file of type T, produces a python COO of type T

2. This is not hard to do, and would merely involve a little extension to `MatrixMarketRead`.

Really, what matters is what you intend (*grin*) for the semantics of the operation.

Reply:
Ok, lets make this strict so only int->int etc. is allowed, but we add a new error type for mismatched reading to make it easier for the user to understand why the reading failed